core/storage/question/gae_models.py

```diff
@@ -244,6 +244,61 @@ def get_question_skill_links_and_skill_descriptions(
         )
         return question_skill_link_models, skill_descriptions, next_cursor_str
 
+    @classmethod
+    def get_question_ids_and_skill_descriptions(
+            cls, question_skill_count, skill_ids, start_cursor):
+        """Fetches question ids and skill descriptions linked to a list of
```

Comment:
Hmm, why are we doing this in the storage layer? If we need the question IDs shouldn't we just use the previous method and deduplicate? (But also, how would you even handle deduplication between pages -- e.g. if the same question ID is mentioned in one page and then again in a different page?)

In any case, there's way too much code duplication here. If we were to keep this method we would need to refactor  so that there isn't repetition. Avoid copy/paste of code in general.

Reply:
Yeah I will get rid of the duplicated code, but just to answer your question here - But also, how would you even handle deduplication between pages -- e.g. if the same question ID is mentioned in one page and then again in a different page?

I don't really understand why there would be same question on different pages. This is what I did: I fetch 30 question_skill_links every time from the backend. Let's say the first page uses 20 of them, then when we go to the second page, it would fetch 30 links again but only shows 21-40 on the frontend. If the first page and second page both use only 10 of them, then it wouldn't fetch from the backend but simply returns data from the frontend. That's how I tried to keep the page size constant.

Reply:
Suppose question Q has skills S1 and S2. The 1st question skill model instance is Q/S1 and the 40th question skill model instance is Q/S2. These will both correspond to the same question ID (but different skills). That would lead to duplicate questions, right?

Reply:
Q/S1 and Q/S2 are always close to each other. In this case, the only situation would be 1: Q/S1, 2: Q/S2, and I handled deduplication already. I am working on making my code more clear. Hopefully you will be able to understand it better later.

Reply:
Also, re the duplicated code: is get_question_skill_links_and_skill_descriptions() still used? If not we should delete it.

Reply:
Got rid of the duplicated code.